lark/load_grammar.py

```diff
@@ -648,17 +649,67 @@ def compile(self, start):
+    This allows them to be compatible even from within zip files.
+    
+    Relative imports are handled, so you can just freely use them.
+    
+    pkg_name: The name of the package. You can probably provide `__name__` most of the time
+    search_paths: All the path that will be search on absolute imports.
+    """
+    def __init__(self, pkg_name, search_paths=("", )):
+        self.pkg_name = pkg_name
+        self.search_paths = search_paths
+        
+    def __repr__(self):
+        return "%s(%r, %r)" % (type(self).__name__, self.pkg_name, self.search_paths)
+
+    def __call__(self, base_paths, grammar_path):
+        if len(base_paths) == 0:
+            to_try = self.search_paths
+        else:
+            assert len(base_paths) == 1
+            if not base_paths[0].startswith('<%s:' % (self.pkg_name,)):
+                # Technically false, but FileNotFound doesn't exist in python2.7, and this message should never reach the end user anyway
+                raise IOError()
+            base_path = base_paths[0].partition(':')[2]
+            if base_path and base_path[0] == '/':
+                base_path = base_path[1:]
+            to_try = [base_path]
+        for path in to_try:
+            full_path = os.path.join(path, grammar_path)
+            text = None
+            with suppress(IOError):
+                text = pkgutil.get_data(self.pkg_name, full_path)
+            if text is None:
+                continue
+            return '<%s:/%s>' % (self.pkg_name, full_path), text.decode()
+        raise IOError()
+
+stdlib_loader = FromPackageLoader('lark', IMPORT_PATHS)
+
 
 _imported_grammars = {}
-def import_grammar(grammar_path, re_, base_paths=[]):
+def import_grammar(grammar_path, re_, base_paths=[], import_sources=[]):
     if grammar_path not in _imported_grammars:
-        import_paths = base_paths + IMPORT_PATHS
-        for import_path in import_paths:
+        # import_sources take priority over base_paths since they should handle relative imports and ignore everthing else.
+        # Question: should the stdlib_loader really be pushed to the end?
+        import_paths = import_sources + base_paths + [stdlib_loader] 
+        for source in import_paths:
+            text = None
             with suppress(IOError):
-                joined_path = os.path.join(import_path, grammar_path)
-                with open(joined_path, encoding='utf8') as f:
-                    text = f.read()
-                grammar = load_grammar(text, joined_path, re_)
+                if callable(source):
+                    joined_path, text = source(base_paths, grammar_path)
+                else:
+                    joined_path = os.path.join(source, grammar_path)
+                    with open(joined_path, encoding='utf8') as f:
+                        text = f.read()
+            if text is not None:
```

Comment:
Why would `text` be None?

Reply:
Because the `with suppress()` statement can fail and then the previous value of text will stay. I assign `None` to it before hand. Alternatively, I can rework it into a `try except else` statement if you want

Reply:
Yes, that's why I loaded the text inside the suppress. But if you think it's better outside (I understand, imports and such), it's better to use `try.. except`

Reply:
Read the next comment.

Reply:
I already said I understand. What did that comment add to *my* comment?

Reply:
Sorry, I misread your comment.